donations/views.py

```diff
@@ -0,0 +1,90 @@
+import json
+import base64
+import requests
+import urlparse
+from django.shortcuts import render
+from django.core.urlresolvers import reverse
+from django.conf import settings
+from django.http import HttpResponse, JsonResponse
+from django.contrib.sites.models import Site
+from django.contrib.auth.models import User
+from django.views.generic import ListView
+from django.views.decorators.csrf import csrf_exempt
+from models import Donation, DonationCampaign
+from utils.mail import send_mail_template
+
+
+@csrf_exempt
+def donation_complete(request):
+    '''This view listens to a notification made from paypal when someone makes
+    a donation, it validates the data and then stores the donation.
+    '''
+    params = {'cmd': '_notify-validate'}
+    for key, value in request.POST.items():
+        params[key] =  value
+    req = requests.post(settings.PAYPAL_VALIDATION_URL, data=params)
+    if req.text == 'VERIFIED':
+        extra_data = json.loads(base64.b64decode(params['custom']))
+        campaign = DonationCampaign.objects.get(id=extra_data['campaign_id'])
+        user = None
+        if 'user_id' in extra_data:
+            user = User.objects.get(id=extra_data['user_id'])
+
+        Donation.objects.get_or_create(transaction_id=params['txn_id'], defaults={
+            'email': params['payer_email'],
+            'display_name': extra_data['name'],
+            'amount': params['mc_gross'],
+            'currency': params['mc_currency'],
+            'user': user,
+            'campaign': campaign})
+
+        send_mail_template(\
+                u'Donation',
+                'donations/email_donation.txt',
+                {'user': user, 'amount': params['mc_gross'], 'display_name': extra_data['name']},
+                None, params['payer_email'])
+    return HttpResponse("OK")
+
+
+def donate(request):
+    ''' Donate page: display form for donations where if user is logged in
+    we give the option to doneate anonymously otherwise we just give the option
+    to enter the name that will be displayed.
+    If request is post we generate the data to send to paypal.
+    '''
+    if request.method == 'POST':
+        campaign = DonationCampaign.objects.order_by('date_start').last()
+        returned_data = {"name": None, "campaign_id": campaign.id}
+        annon = request.POST.get('annon', None)
```

Comment:
couldn't all of this stuff be done by creating a Form object?

it might be more work, but probably (much) less error-prone...